### PR TITLE
Make DTOUtil stop relying on exceptions

### LINKED ISSUE #392: Converter's DTOUtil is overly reliant on exceptions for normal execution flow
DTOUtil.isDTOType() [1] relies on exceptions to find out whether a class has a public constructor and to find out if it has any methods - as a DTO should have a public no-args constructor and no methods beyond the ones defined by Object.class. 

This works but is wasteful and slow. We should implement a different strategy to identify a DTO class as such.

[1] https://github.com/osgi/osgi/blob/main/org.osgi.util.converter/src/org/osgi/util/converter/DTOUtil.java

### PR DESCRIPTION
**Pull request by bosschaert**
Fixes #392

### CODE REVIEW

**Review by karlpauls — APPROVED** [review]
LGTM - thanks!

**Comment by bosschaert** [comment]
I force-pushed the commit now with the signoff and my registered email address.

File: org.osgi.util.converter/src/org/osgi/util/converter/DTOUtil.java [hunk 1]
```diff
@@ -18,45 +18,53 @@
...
-			} catch (NoSuchMethodException | SecurityException e) {
-				// No public zero-arg constructor, not a DTO
+			boolean hasPublicNoargsCtor = false;
+			for (Constructor< ? > ctor : cls.getConstructors()) {
+				if (ctor.getParameterCount() == 0) {
+					hasPublicNoargsCtor = true;
+					break;
+				}
+			}
+			if (!hasPublicNoargsCtor) {
```
Comment (bjhargrave) [inline org.osgi.util.converter/src/org/osgi/util/converter/DTOUtil.java:47]:
```suggestion
			if (Arrays.stream(cls.getConstructors())
					.noneMatch(ctor -> ctor.getParameterCount() == 0)) {
				// No public zero-arg constructor, not a DTO
```
More concise alternative.

File: org.osgi.util.converter/src/org/osgi/util/converter/DTOUtil.java [hunk 2]
```diff
@@ -18,45 +18,53 @@
...
+			for (Method om : OBJECT_CLASS_METHODS) {
+				if (om.getName().equals(m.getName())) {
+					if (Arrays.equals(om.getParameterTypes(),
+							m.getParameterTypes())) {
+						objectClassMethod = true;
+						break;
+					}
+				}
+			}
+			if (!objectClassMethod) {
```
Comment (bjhargrave) [inline org.osgi.util.converter/src/org/osgi/util/converter/DTOUtil.java:63]:
```suggestion
			if (Arrays.stream(OBJECT_CLASS_METHODS)
					.noneMatch(om -> om.getName().equals(m.getName())
							&& Arrays.equals(om.getParameterTypes(),
									m.getParameterTypes()))) {
				// Not a method defined by Object.class (or override of such
				// method)
```

More concise alternative.

**Comment by bosschaert** [comment]
Sure we can use a stream based variant too. Thanks @bjhargrave !

**Review by bjhargrave — APPROVED** [review]

**Pull request merged**